enhancements/installer/aws-custom-region-and-endpoints.md

```diff
@@ -0,0 +1,321 @@
+
+## Motivation
+
+### Goals
+
+1. Allow users to specify API endpoints for some or all required AWS services.
+
+2. Provide users capability to install in custom regions on AWS.
+
+3. Provide automated handling for most users with respect to getting AMI available in custom regions.
+
+### Non-Goals
+
+1. There are no plans to allow self-signed server certificates for endpoints specifically.
+
+2. Allowing users to use un-encrypted AMIs for the ec2 instances.
+
+## Proposal
+
+Currently the installer only allows users to specify the regions that has RHCOS published AMIs. But we would support regions that do not have RHCOS AMI published in 2 ways:
+
+1. Copying AMIs from us-east-1 to the region where the user is installing.
+Currently the installer only copies and encrypts the AMI published by RHCOS in the same region for low latency. But the installer should be able to support copying and encrypting the AMI published by RHCOS for us-east-1 region to the region user is installing the cluster. This will cause the time to copy to the AMI to that region, but it should enable users to user these new regions.
+
+2. Allowing users to provide the AMI in the install-config for regions that cannot import AMIs from other regions like US Gov Cloud.
+This is already supported by [install-config][install-config-aws-ami]. The users will have to use the user-provided-installation workflow to get the AMI imported to the region where the cluster will be installed.
+The installer will use the AMI provided and copy-encrypt to an AMI for the cluster, keeping the behavior consistent with other workflows.
+
+To support custom endpoints for AWS APIs, the install-config would allow the users to provide a list of endpoints for the services. The endpoints will be used by the installer to call AWS APIs to perform validation, create the MachineSet objects, and will be used to configure the terraform AWS provider. When the custom endpoints will be set, the installer will validate that the endpoints are reachable. The installer requires AWS API endpoints for EC2, IAM, Route53, ELB, ELBv2, STS, but the installer will allow the user to provide API endpoints for a subset of the endpoints under the assumption that default API endpoints are usable.
+
+When custom endpoints are provide for AWS APIs, the cluster operators also need the discover the information, therefore, the installer will set the AWS API endpoints in the `spec` of the `infrastructures.config.openshift.io` `cluster` object.
+
+### User Stories
+
+#### Using private VPC endpoints
+
+#### Installing to US Gov Cloud
+
+#### Using FIPS endpoints for US Gov Cloud
+
+#### Installing to custom region
+
+#### Installing to new regions without native RHCOS support in the region
```

Comment:
This item isn't just relevant for RHCOS support - components like the image registry need their SDKs upgraded as new regions are added.

In the case of the image registry, we have a `regionEndpoint` override in our operator config which lets admins override the S3 endpoint the registry talks to. In implementing this enhancement, the registry operator would likely read the `CustomEndpoint` for S3 and apply it to the `regionEndpoint` override. In theory customers can add the new region endpoints as `CustomEndpoint` values, and the operator would take care of the rest.

See https://bugzilla.redhat.com/show_bug.cgi?id=1787604